Kernel/LPOComparator.cpp

```diff
@@ -52,438 +48,193 @@ bool unify(TermList tl1, TermList tl2, TermList& orig1, TermList& orig2)
+            goto loop_end;
           }
-          break;
-        }
-
-        // Otherwise, we have to create a branching point here.
-
-        // Replace the previous 'equal' instructions with a jump to here.
-        if (prevStartIndex != INDEX_UNINITIALIZED) {
-          updateBranchInRange(res->first, prevStartIndex, prevEndIndex,
-            Branch::eq(), Branch::jump(res->first.size()));
-        }
-        prevStartIndex = res->first.size();
-        prevEndIndex = res->first.size() + compRes->first.size();
-        // The majo branch will be after the comparison instructions
-        Branch majoBranch{
-          majoRes.second,
-          (uint16_t)(majoRes.second == BranchTag::T_JUMP ? res->first.size() + compRes->first.size() : 0)
-        };
-        // The alpha branch will be after the majo branch instructions
-        Branch alphaBranch{
-          alphaRes.second,
-          (uint16_t)(alphaRes.second == BranchTag::T_JUMP ? res->first.size() + compRes->first.size() + majoRes.first.size() : 0)
-        };
-
-        // push all three sets of instructions if needed
-        pushInstructions(res->first, compRes->first, Branch::eq(), majoBranch, alphaBranch);
-
-        if (majoRes.second == BranchTag::T_JUMP) {
-          pushInstructions(res->first, majoRes.first, Branch::eq(), Branch::gt(), Branch::inc());
-        }
-
-        if (alphaRes.second == BranchTag::T_JUMP) {
-          pushInstructions(res->first, alphaRes.first, Branch::eq(), Branch::gt(), Branch::inc());
-        }
-
-        if (!unify(s_arg,t_arg,tl1s,tl2s)) {
-          // If we cannot unify, the rest of the iterations will be
-          // 'incomparable', update the previous 'equal' values to that.
-          updateBranchInRange(res->first, prevStartIndex, prevEndIndex, Branch::eq(), Branch::inc());
-          break;
         }
+        *curr = node->eqBranch;
+        break;
       }
-      break;
-    }
-    case Ordering::GREATER: {
-      auto subres = majoChain(lpo, tl1, t, 0);
-      if (subres.second == BranchTag::T_JUMP) {
-        pushInstructions(res->first, subres.first, Branch::eq(), Branch::gt(), Branch::inc());
-      } else {
-        res->second = subres.second;
+      case Ordering::GREATER: {
+        ASS(t->arity());
+        majoChain(&branch, node->lhs, t, 0, node->gtBranch, node->incBranch);
+        break;
       }
-      break;
-    }
-    case Ordering::LESS: {
-      auto subres = alphaChain(lpo, s, 0, tl2);
-      if (subres.second == BranchTag::T_JUMP) {
-        pushInstructions(res->first, subres.first, Branch::eq(), Branch::gt(), Branch::inc());
-      } else {
-        res->second = subres.second;
+      case Ordering::LESS: {
+        ASS(s->arity());
+        alphaChain(&branch, s, 0, node->rhs, node->gtBranch, node->incBranch);
+        break;
+      }
+      default:
+        ASSERTION_VIOLATION;
       }
-      break;
     }
-    default:
-      ASSERTION_VIOLATION;
+loop_end:
+    node->release();
```

Comment:
This looks RAIIish (i.e. should be a smart pointer of some kind)...is it, or did I miss the path through the loop where you don't release the `node`?

Reply:
We could actually just put it into a `Branch` (or whatever the RAII container ends up being, e.g. `shared_ptr`). The point here is that the original container is disrupted so `node` could be deleted while we're processing it.